staging/src/k8s.io/apiserver/pkg/storage/etcd3/feature_support_checker.go

```diff
@@ -0,0 +1,164 @@
+	client *clientv3.Client
+	// mutext protects featureSupportCache variable.
+	mutex sync.RWMutex
+	// featureSupportCache store result of feature.
+	featureSupportCache map[storage.Feature]featureSupportInfo
+}
+
+func newDefaultFeatureSupportChecker(client *clientv3.Client) *DefaultFeatureSupportChecker {
+	return &DefaultFeatureSupportChecker{
+		client:              client,
+		featureSupportCache: make(map[storage.Feature]featureSupportInfo),
+	}
+}
+
+// Supports checks whether a specific feature is supported by etcd.
+// TODO: Extend this implementation to support feature-checking for the cacher.
+func (c *DefaultFeatureSupportChecker) Supports(ctx context.Context, feature storage.Feature) (bool, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	// Check if the feature is already cached
+	if info, ok := c.featureSupportCache[feature]; ok {
+		return info.Supported, info.Err
+	}
+
+	// Feature-specific support check
+	supported, err := c.checkFeatureSupport(ctx, feature)
+
+	// Cache the result
+	c.featureSupportCache[feature] = featureSupportInfo{
+		Supported: supported,
+		Err:       err,
+	}
+
+	return supported, err
+}
+
+// checkFeatureSupport checks the support of a specific feature.
+func (c *DefaultFeatureSupportChecker) checkFeatureSupport(ctx context.Context, feature storage.Feature) (bool, error) {
+	if feature == RequestWatchProgress {
+		return isDeprecated(ctx, c.client)
+	}
+	return true, nil // Unsupported features are assumed to be supported
+}
+
+// checkVersionEtcdClientInterface defines the interface required for testing purposes in IsDeprecated function of the interface.
+// It is intended to be used exclusively within the IsDeprecated function for interacting with etcd client components.
+type checkVersionEtcdClientInterface interface {
+	// Endpoints returns the list of endpoints for the etcd client.
+	Endpoints() []string
+
+	// Status retrieves the status information from the etcd client connected to the specified endpoint.
+	// It takes a context.Context parameter for cancellation or timeout control.
+	// It returns a clientv3.StatusResponse containing the status information or an error if the operation fails.
+	Status(ctx context.Context, endpoint string) (*clientv3.StatusResponse, error)
+}
+
+// isDeprecated evaluates whether the current version of etcd is deprecated based on known issues.
+// It checks if the version of etcd is suitable for a specific feature by considering documented issues, such as:
+//   - Issue #15220: https://github.com/etcd-io/etcd/issues/15220
+//   - KEP 2340: Consistent reads from cache: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md#bug-in-etcd-progress-notification
+//
+// It returns a boolean indicating whether the version is deprecated and an error if any occurred during the version check.
+func isDeprecated(ctx context.Context, c checkVersionEtcdClientInterface) (bool, error) {
```

Comment:
This no longer is just check deprecation, you check for support of RequestWatchProgress. The name of the function should reflect that.

Reply:
Thanks for pointing this out, yep so many changed i forgot about this. 
What about `isSupportedRequestWatchProgress` ?

Reply:
Updated, please take a look.